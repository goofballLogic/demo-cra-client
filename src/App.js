import React from 'react';
import { AppStoreProvider } from "./AppStore";
import Body from "./Body.js";
import useStore from "./hooks/useStore";
import useAPIdata from './hooks/useAPIdata';

const initialState = { posts: {}, comments: {}, profile: {} };

function App() {

  const [store, dispatch] = useStore(initialState);
  useAPIdata(dispatch);

  return (
    <AppStoreProvider value={store}>
      <Body />
    </AppStoreProvider>
  );

}

export default App;
