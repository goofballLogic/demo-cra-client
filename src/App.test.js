import React from 'react';
import App from './App';
import waitUntil from 'async-wait-until';
import renderer, { act } from 'react-test-renderer';

const fakePosts = [
  { "id": 1, "title": "Things are great" },
  { "id": 2, "title": "It's all terrible" },
  { "id": 3, "title": "I'm starting to learn things" }
];

const fakeComments = [
  { "id": 1, "body": "This post is rather misguided in my opinion", "postId": 1 },
  { "id": 2, "body": "Great post!", "postId": 1 }
];

const fakeProfile = {
  "name": "andrew"
};

const ok = json => ({ ok: true, status: 200, json: () => json });
const notFound = () => ({ ok: false, status: 404 });

describe("given APIs return valid JSON", () => {

  let calls = [];
  beforeEach(() => {
    jest.spyOn(global, "fetch").mockImplementation(url => {
      calls.push(url);
      if (url.endsWith("/posts")) return ok(fakePosts);
      if (url.endsWith("/comments")) return ok(fakeComments);
      if (url.endsWith("/profile")) return ok(fakeProfile);
      return notFound();
    });
  });

  describe("when i render the app and wait for data to load", () => {

    let tree;

    beforeEach(async () => {

      const app = <App />;
      act(() => {
        tree = renderer.create(app);
      });
      await waitUntil(() => calls.length > 2);
      act(() => {

        tree.update(app);

      });

    });

    it("should have done stuff", () => {

      expect(tree.toJSON()).toMatchSnapshot();

    });

  });

  afterEach(() => global.fetch.mockClear());
});