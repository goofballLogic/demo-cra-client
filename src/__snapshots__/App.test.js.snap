// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given APIs return valid JSON when i render the app and wait for data to load should have done stuff 1`] = `
<main>
  <section
    className="posts"
  >
    <article
      className="post"
    >
      <h1>
        Things are great
      </h1>
      <p>
        Some text
      </p>
      <p
        className="comment"
      >
        This post is rather misguided in my opinion
      </p>
      <p
        className="comment"
      >
        Great post!
      </p>
    </article>
    <article
      className="post"
    >
      <h1>
        It's all terrible
      </h1>
      <p>
        Some text
      </p>
      No comments yet
    </article>
    <article
      className="post"
    >
      <h1>
        I'm starting to learn things
      </h1>
      <p>
        Some text
      </p>
      No comments yet
    </article>
  </section>
  <article
    className="profile"
  >
    Profile
    <div>
      <h1>
        andrew
      </h1>
    </div>
  </article>
</main>
`;
